### PR TITLE
Chef 14 compatibility

### PR DESCRIPTION
**Pull request by ledgr**
### Description

[Chef 12 is deprecated as of 2018 April](https://www.chef.io/eol-chef12-and-chefdk1/). This PR updates dovecot-cookbook to be compatible with Chef 14

### Issues Resolved

- [x] Remove old dovecot Ohai plugin version
- [x] Remove gem version locks in metadata.rb
- [x] Update platform versions and review old ones (preferably supported by Fauxhai)
- [x] Update tests after upgrading
- [x] Update Travis integration tests matrix

### Contribution Check List

- [X] All tests pass.
- [X] New functionality includes testing.
- [X] New functionality has been documented in the README and metadata if applicable.

See [CONTRIBUTING.md](https://github.com/zuazo/dovecot-cookbook/blob/master/CONTRIBUTING.md).

### NOTES

* Ohai dependency will have to be removed for Chef 15 to work.
* Service provider conditional selection is moved from attributes to recipe/service.rb.
   Keeping condition in attributes return Class object instead of class path making
   serverspec tests to fail

### CODE REVIEW

**Comment by ledgr** [comment]
Fixes: https://github.com/zuazo/dovecot-cookbook/issues/30

**Comment by ledgr** [comment]
Compatible with Chef 12 and Chef 13.

**Comment by ledgr** [comment]
Okay, I'm really not sure why it reports:

```
Untouched Resources:

  directory[/etc/dovecot/.]          dovecot/recipes/conf_files.rb:32
``` 

If test passes:

```
.  avoids the creation of /etc/dovecot/. (with dot) directory
```

I might take a look into it after a few days.

File: test/unit/recipes/default_spec.rb [hunk 1]
```diff
@@ -24,20 +22,20 @@
...
-  ).each do |recipe|
+  ].each do |recipe|
     it "includes #{recipe} recipe" do
       expect(chef_run).to include_recipe(recipe)
     end
   end # each recipe
 
   xcontext 'with install from source (not yet implemented)' do
-    before { node.set['dovecot']['install_from'] = 'source' }
+    before { node.normal['dovecot']['install_from'] = 'source' }
```
Comment (lamont-granquist) [inline test/unit/recipes/default_spec.rb]: even though these are just tests, node.normal shouldn't really be used here, it should most likely just be node.default, or node.override if it has to be that way...

Comment (ledgr) [reply test/unit/recipes/default_spec.rb]: I just did a straightforward replace as per deprecation documentation here: https://chef.github.io/chef-rfc/rfc082-deprecation-policy.html

It says `node.set` must be replaced by `node.normal` which is an alias.

I will evaluate `node.default`

Comment (lamont-granquist) [reply test/unit/recipes/default_spec.rb]: yeah i wrote that deprecation =)

the "We highly suggest you evaluate your usage of node.set / node.normal though to see if that is actually the attribute behavior you desire" bit probably needs to be rewritten a little less passively and more aggressively....  that's sort of the point which is to stop using the normal level accidentally and convert to default/override...

**Comment by zuazo** [comment]
Wow! Amazing work @ledgr! Thank you very much!

Give me a couple of days to review it. I also need to check the other PR about ohai. On Sunday I'm gonna start reviewing all those pending PRs.

File: Gemfile [hunk 2]
```diff
@@ -12,53 +8,52 @@ source 'https://rubygems.org'
...
+  gem 'berkshelf', '~> 7.0'
   gem 'rake', '~> 12.0'
-  gem 'berkshelf', '~> 6.0'
 end
 
 group :style do
-  gem 'foodcritic', '~> 10.2.0'
-  gem 'rubocop', '~> 0.39.0'
+  gem 'foodcritic'
+  gem 'rubocop'
```
Comment (zuazo) [inline Gemfile]: Maybe I wrong, but I think removing the versions from here will make its maintainability more difficult. Mainly due to big rule changes between different foodcritic and rubocop versions. So I prefer to adhere to some specific versions.

File: Gemfile [hunk 3]
```diff
@@ -12,53 +8,52 @@ source 'https://rubygems.org'
...
 end
 
 group :unit do
   gem 'chef', chef_version unless chef_version.nil?
-  gem 'chefspec', '~> 6.0'
-  gem 'simplecov', '~> 0.13.0'
-  gem 'should_not', '~> 1.1'
+  gem 'chefspec'
+  gem 'should_not'
+  gem 'simplecov'
```
Comment (zuazo) [inline Gemfile]: These are not as important as the previous ones, but I don't know why did you remove the version constraints here too. Is this some new pattern of the community that I'm not aware of?

File: attributes/conf_10_ssl.rb [hunk 4]
```diff
@@ -52,10 +46,11 @@
       default['dovecot']['conf']['ssl_key'] = nil
     else
       default['dovecot']['conf']['ssl_cert'] = '</etc/dovecot/dovecot.pem'
-      default['dovecot']['conf']['ssl_key'] =
-        '</etc/dovecot/private/dovecot.pem'
+      default['dovecot']['conf']['ssl_key'] = '</etc/dovecot/private/dovecot.pem'
     end
   end
+when 'suse'
+  default['dovecot']['conf']['ssl'] = true if node['platform_version'].to_i < 13
```
Comment (zuazo) [inline attributes/conf_10_ssl.rb]: :+1:

File: recipes/conf_files.rb [hunk 5]
```diff
@@ -32,7 +30,6 @@
   end.uniq
 conf_files_dirs.each do |dir|
   directory "#{node['dovecot']['conf_path']}/#{dir}" do
-    name ::File.join(node['dovecot']['conf_path'], dir)
```
Comment (zuazo) [inline recipes/conf_files.rb]: The `File.join` is there for different OS compatibility. But yeah, probably should be removed :+1:

File: recipes/ohai_plugin.rb [hunk 6]
```diff
@@ -19,21 +17,15 @@
...
   action :nothing
 end
 
 ohai_plugin 'dovecot' do
-  name 'dovecot'
-  source_file "#{source_dir}/dovecot.rb.erb"
+  source_file 'ohai_plugins/dovecot.rb.erb'
   resource :template
-  variables enable_build_options: ohai_build_options
+  variables enable_build_options: build_options
```
Comment (zuazo) [inline recipes/ohai_plugin.rb]: Good!

File: test/unit/recipes/default_spec.rb (hunk 1 above)
Comment (zuazo) [reply test/unit/recipes/default_spec.rb]: Probably this should be `override`, yes.

File: recipes/service.rb [hunk 7]
```diff
@@ -23,6 +21,12 @@
...
-  provider node['dovecot']['service']['provider']
-  action [:enable, :start]
+  if node['platform'] == 'ubuntu' &&
+     Gem::Requirement.new(['>= 13.10', '<= 15.10'])
+                     .satisfied_by?(Gem::Version.new(node['platform_version']))
+    provider Chef::Provider::Service::Upstart
+  elsif node['platform'] == 'debian' && node['platform_version'].to_i < 9
+    provider Chef::Provider::Service::Debian
+  end
+  action %i[enable start]
```
Comment (zuazo) [inline recipes/service.rb]: I see you moved the `node['dovecot']['service']['provider']` attribute logic from attributes file to the recipe. Is this a requirement of new Chef versions or just to avoid having code in attribute files? Previously we can override the provider, now we cannot.

Anyway, doing this change, will require to update the [Distribution Service Configuration](https://github.com/zuazo/dovecot-cookbook#distribution-service-configuration) section of the README. Probably other sections like [Supported Platforms](https://github.com/zuazo/dovecot-cookbook#supported-platforms) will also need to be updated.

**Review by zuazo — CHANGES_REQUESTED** [review]
Thank you so much for your hard work! :smiley: I added some comments to your changes. Please if you could check them.

**Comment by ledgr** [comment]
@zuazo @lamont-granquist I have changed attribute precedence levels in tests, and I have put back version locks in Gemfile.

File: recipes/service.rb (hunk 7 above)
Comment (zuazo) [reply recipes/service.rb]: Oh, I did not see that comment. My fault.

**Review by zuazo — APPROVED** [review]
Sorry for the delay. Merging this. Thanks for your hard work :+1:

**Comment by ledgr** [comment]
🎉 thank you 😄
